compiler/pipes/calc-bad-vars.cpp

```diff
@@ -118,22 +119,80 @@ struct FuncCallGraph {
 class CalcBadVars {
 private:
   class MergeBadVarsCallback : public MergeReachalbeCallback<FunctionPtr> {
-    IdMap<vector<VarPtr>> tmp_vars;
+    IdMap<vector<VarPtr>> modified_global_vars;
   public:
-    explicit MergeBadVarsCallback(IdMap<vector<VarPtr>> tmp_vars) : tmp_vars(std::move(tmp_vars)) {}
+    explicit MergeBadVarsCallback(IdMap<vector<VarPtr>> &&modified_global_vars) :
+      modified_global_vars(std::move(modified_global_vars)) {}
+
+    // here we calculate "bad vars" for a function — they will be used in check-ub.cpp
+    // "bad vars" are modified non-primitive globals, that can lead to a potential ub when modified and accessed in subcalls
+    // this method just unions all edges' bad_vars to the current function
+    // this works in a single thread and uses some heuristic for speedup, but generally it means
+    // "for f in component { for e in edge { f.bad_vars += e.bad_vars } }"
     void for_component(const vector<FunctionPtr> &component, const vector<FunctionPtr> &edges) override {
-      std::unordered_set<VarPtr> bad_vars_uniq;
-
+      // optimization 1: lots of functions don't modify globals at all
+      // then we leave f->bad_vars nullptr
+      bool empty = true;
       for (FunctionPtr f : component) {
-        bad_vars_uniq.insert(tmp_vars[f].begin(), tmp_vars[f].end());
+        empty &= modified_global_vars[f].empty();
+      }
+      for (FunctionPtr f : edges) {
+        empty &= f->bad_vars == nullptr;
+      }
```

Comment:
You can shorten the code above, on top of that it wouldn't iterate over edges if not empty `modified_global_vars` has been found
```
bool empty = vk::all_of(component, [this](auto f) { return modified_global_vars[f].empty(); })
          && vk::all_of(edges, [](auto f) { return f->bad_vars == nullptr; });
```

Reply:
ok